test/replaytests/replay_test.go

```diff
@@ -81,92 +84,110 @@ func TestGreetingsWorkflow3(t *testing.T) {
+// Branch workflow normal history file is replayed against modified workflow code which
+// has 2 branches only.  This causes nondetereministic error.
 func TestBranchWorkflowWithExtraBranch(t *testing.T) {
 	replayer := worker.NewWorkflowReplayer()
-
 	replayer.RegisterWorkflowWithOptions(sampleBranchWorkflow2, workflow.RegisterOptions{Name: "branch"})
-
 	err := replayer.ReplayWorkflowHistoryFromJSONFile(zaptest.NewLogger(t), "branch.json")
-	assert.ErrorContains(t, err, "nondeterministic workflow")
+	assert.ErrorContains(t, err, "nondeterministic workflow: missing replay decision")
 }
 
-func TestParallel(t *testing.T) {
+// TestSequentialStepsWorkflow replays a history with 2 sequential non overlapping activity calls (one completes before the other is scheduled)
+// and runs it against new version of the workflow code which only calls 1 activity.
+// This is considered as non-determinism error.
+func TestSequentialStepsWorkflow(t *testing.T) {
 	replayer := worker.NewWorkflowReplayer()
+	replayer.RegisterWorkflowWithOptions(replayerHelloWorldWorkflow, workflow.RegisterOptions{Name: "fx.ReplayerHelloWorldWorkflow"})
+	replayer.RegisterActivityWithOptions(replayerHelloWorldActivity, activity.RegisterOptions{Name: "replayerhello"})
+	err := replayer.ReplayWorkflowHistoryFromJSONFile(zaptest.NewLogger(t), "sequential.json")
+	assert.ErrorContains(t, err, "nondeterministic workflow: missing replay decision")
+}
 
+// Runs simpleParallelWorkflow which starts two workflow.Go routines that executes 1 and 2 activities respectively.
+func TestSimpleParallelWorkflow(t *testing.T) {
+	replayer := worker.NewWorkflowReplayer()
 	replayer.RegisterWorkflowWithOptions(sampleParallelWorkflow, workflow.RegisterOptions{Name: "branch2"})
-
 	err := replayer.ReplayWorkflowHistoryFromJSONFile(zaptest.NewLogger(t), "branch2.json")
 	require.NoError(t, err)
 }
 
-// Should have failed since the first go routine has only one branch whereas the history has two branches.
-// The replayer totally misses this change.
-func TestParallel2(t *testing.T) {
+// Runs modified version of simpleParallelWorkflow which starts 1 less activity in the second workflow-gouroutine.
+// This is considered as non-determinism error.
+func TestSimpleParallelWorkflowWithMissingActivityCall(t *testing.T) {
 	replayer := worker.NewWorkflowReplayer()
-
 	replayer.RegisterWorkflowWithOptions(sampleParallelWorkflow2, workflow.RegisterOptions{Name: "branch2"})
-
 	err := replayer.ReplayWorkflowHistoryFromJSONFile(zaptest.NewLogger(t), "branch2.json")
-	require.NoError(t, err)
+	assert.ErrorContains(t, err, "nondeterministic workflow: missing replay decision")
+}
+
+// Runs a history which ends with WorkflowExecutionContinuedAsNew. Replay fails because of the additional checks done
+// for continue as new case by replayWorkflowHistory().
+// This should not have any error because it's a valid continue as new case.
+func TestContinueAsNew(t *testing.T) {
```

Comment:
this case will be handled in a follow up PR

Reply:
yea, replayer shouldn't care.  _shadower_ might, but that's completely different.

separately, now that I look at this... it feels strange that it doesn't return `future, err`.  replaying to get the result (from code) seems entirely reasonable, no need to only return error...